/*
 ** cdn/broker message definitions
 */

package CDN;
//syntax = "proto2";

enum constants {
  CDN_PROTO_VERSION = 1;
}

message Request {
  optional uint32 bandwidth = 1; // current video bandwidth @ client
  optional uint32 tput      = 2; // currenty client video throughput
  optional string last_cdn  = 3; // cdn client is currently using
}

message Reply {
  optional string cluster = 1;
  optional string ad      = 2;
  optional string hid     = 3;
}

message Cluster {
  required string name = 1;	// hostname of the 'cluster'
  optional uint32 rtt  = 2;	// avg ping rtt time in ms. Infinate if blank
  optional uint32 loss = 3;	// packet loss percentage 0-100
}

message Scores {
  repeated Cluster clusters = 1; // list of scores for connections to CDN clusters
}

enum msg_type {
  CDN_REQUEST_MSG  = 1;
  PING_SCORES_MSG  = 2;
  CDN_REPLY_MSG    = 3;
}

message CDNMsg {
  required uint32   version = 1;
  required msg_type type    = 2;
  optional string   client  = 3; // client requesting a cdn

  oneof payload {
    Request request = 4;
    Scores  scores  = 5;
    Reply   reply   = 6;
  }
}

